onnxscript/function_libs/torch_lib/ops/nn.py

```diff
@@ -2076,6 +2076,11 @@ def _aten_scaled_dot_product_attention_bool_mask_onnx(
         op.Add(op.MatMul(query_scaled, key_transposed_scaled), attn_mask),
         axis=-1,
     )
+    # When using scaled dot product attention with a boolean mask, the softmax operation might return NaN values
+    # due to the presence of -inf in an entire row (padding tokens), resulting in 0/0 (NaN) in the softmax output.
+    # This is because there's no safe/masked softmax imp in ONNX, so we need to handle NaN values explicitly to match
+    # the behavior of PyTorch with boolean masks.
+    attn_weight = op.Where(op.IsNaN(attn_weight), zero, attn_weight)
```

Comment:
Thanks ... overall, this looks reasonable. Just wondering: is there a specification of this behavior in pytorch? I am also curious under what circumstances this arises in practice? If one is not paying attention to any of the key/values, that seems odd.

Reply:
> If one is not paying attention to any of the key/values, that seems odd.

this happens when sequences are left padded, which is what's done when trying to do batched text generation.

> is there a specification of this behavior in pytorch?

it is discussed in depth in this issue https://github.com/pytorch/pytorch/issues/103749